# (12주차) 11월30일
> 기울기소멸, 추천시스템 

- toc:true
- branch: master
- badges: true
- comments: true
- author: 최규빈

## 강의영상 

> youtube: https://youtube.com/playlist?list=PLQqh36zP38-wXD4q0NOEMsrnCM1eQTx18

`-` (1/3) 기울기소멸 

`-` (2/3) 추천시스템

`-` (3/3) 숙제설명 

## 기울기소멸 

### 고요속의 외침 

`-` https://www.youtube.com/watch?v=ouitOnaDtFY

`-` 중간에 한명이라도 잘못 말한다면.. 

### 정의 

`-` In machine learning, the vanishing gradient problem is encountered when training artificial neural networks with gradient-based learning methods and backpropagation. 

### 이해

`-` 당연한것 아닌가? 
- 그레디언트 기반의 학습 (그레디언트 기반의 옵티마이저): 손실함수의 기울기를 통하여 업데이트 하는 방식 
- 역전파: 손실함수의 기울기를 구하는 테크닉 (체인룰 + $\alpha$). 구체적으로는 (1) 손실함수를 여러단계로 쪼개고 (2) 각 단계의 미분값을 각각 구하고 (3) 그것들을 모두 곱하여 기울기를 계산한다. 
- 0 근처의 숫자를 계속 곱하면 터지거나 0으로 간다. (사실 안정적인 기울기가 나올 것이라고 생각하는것 자체가 사실 이상함, 어떤 숫자 하나라도 0에 아주 가까우면 결국 곱한 결과는 0에 가까워짐) 

In [2]:
import numpy as np 

In [3]:
grads = np.random.uniform(low=-2,high=2,size=100) 
grads

array([-3.76493375e-01,  6.65719034e-01,  1.73260538e+00,  1.30041504e-01,
       -1.62214107e+00, -1.93153336e+00, -5.04508041e-01, -4.59116660e-01,
        1.33384020e-03, -1.64406851e+00,  1.07471196e-01, -8.21608971e-01,
        1.85564465e+00, -1.84471611e+00, -1.32348592e+00,  9.76030199e-01,
        1.13838097e+00,  1.60923664e+00,  3.09191133e-01,  2.95771780e-01,
       -6.27697464e-01, -8.60096935e-01, -6.25109515e-01, -9.78694769e-01,
        2.54836124e-01, -1.33566531e-01,  6.24202832e-01,  5.13782994e-01,
       -1.35709910e+00,  4.18476984e-01,  1.91613878e-01, -1.25936367e+00,
        1.28501653e+00, -7.54417286e-01,  1.47858953e+00,  6.01868175e-01,
        7.61273437e-01, -1.88316907e+00, -1.48881876e+00, -1.67797205e+00,
        2.90899639e-01,  2.98909276e-01,  1.57619503e+00, -2.70708045e-01,
       -1.88806940e+00, -6.46244104e-01, -5.08792551e-01, -1.84687770e+00,
       -1.02348090e+00, -1.07099434e+00, -4.68827191e-01,  1.12350420e+00,
        1.74882624e+00, -

In [4]:
grads.prod()

1.6076914063193976e-16

- 기울기가 소멸함 

In [5]:
grads = np.random.uniform(low=-5,high=5,size=100) 
grads.prod()

-3.1645626270792544e+25

- 기울기가 폭발함. 

In [6]:
grads = np.random.uniform(low=-1,high=3.5,size=100) 
grads.prod()

56.379313631589255

`-` 도깨비: 기울기가 소멸하기도 하고 터지기도 한다. 

### 해결책 (기울기 소멸에 대한 해결책) 

> Note: 잘 정리되어있는것이 없어서 제 머리속에서 정리했습니다. 부정확할수도 있어요 

`-` 개념 
- 데이터 $\to$ (아키텍처,손실함수,역전파,업데이트) 

`-` Multi-level hierarchy
- 여러층을 쪼개서 학습하자 $\to$ 어떻게? 사전학습, 층별학습 
- 기울기소실문제를 해결하여 딥러닝을 유행시킨 태초의(?) 방법임. 
- 결국 입력자료를 바꾼뒤에 학습하는 형태 

`-` Faster hardware
- GPU를 중심으로 한 테크닉 
- 근본적인 문제해결책은 아니라는 힌튼의 비판 
- CPU를 쓸때보다 GPU를 쓰면 약간 더 깊은 모형을 학습할 수 있다 정도? 
- 교수님은 역전파 부분 즉, 하드웨어 부분을 건드리는 방법이라고 생각

`-` Residual networks
- 훌륭한 접근법중 하나임 
- 아키텍처를 변경하는 방법이지만, 사실상 손실함수를 부드럽게 만드는 기법으로 이해해도 된다. 
- 솟컷이라는 아키텍처를 추가하여 이리저리 실험해보니까 손실함수가 부드러워졌다.  <--- 이런게 아니고 
- 손실함수를 부드럽게 하기 위해서는 층별의차이(residual)를 학습하는게 유리할 것 같다. 그런데 이것을 위한 효과를 주기 위해서는 단지 아키텍처에 숏만만 추가하면 되겠다. <--- 이런 모티브였을 것이다. 
- 교수님은 손실함수 부분을 건드린 것으로 이해
- 이로 인해 resnet이 나오게 됨

`-` Other activation functions
- 렐루의 개발

`-` 배치정규화 
- 어쩌다보니 되는것. 
- 배치정규화는 원래 공변량 쉬프트를 잡기 위한 방법임. 그런데 기울기 소멸에도 효과가 있음. 현재는 기울기소멸문제에 대한 해결책으로 빠짐없이 언급되고 있음. 2015년의 원래 논문에는 기울기소멸에 대한 언급은 없었음. (https://arxiv.org/pdf/1502.03167.pdf)
- 심지어 배치정규화는 오버피팅을 잡는효과도 있음 (이것은 논문에 언급했음) 

`-` **기울기를 안구하면 안되나?**`
- 베이지안 최적화기법: (https://arxiv.org/pdf/1807.02811.pdf) $\to$ GPU를 어떻게 쓰지? $\to$ 느리다 

## 추천시스템 

### import 

In [ ]:
!pip install --upgrade fastai

In [1]:
import pandas as pd 
import torch 
from fastai.collab import * 
from fastai.tabular.all import * 

### Data (시뮬레이션) 

In [3]:
df=pd.read_csv('https://raw.githubusercontent.com/guebin/2021BDA/master/_notebooks/2021-11-30-recommend.csv')
df

user  item    rating item_name
0       1    15  1.084308       홍차5
1       1     1  4.149209       커피1
2       1    11  1.142659       홍차1
3       1     5  4.033415       커피5
4       1     4  4.078139       커피4
..    ...   ...       ...       ...
995   100    18  4.104276       홍차8
996   100    17  4.164773       홍차7
997   100    14  4.026915       홍차4
998   100     4  0.838720       커피4
999   100     7  1.094826       커피7

[1000 rows x 4 columns]

### user-item matrix 

`-` 아래와 같은 매트릭스를 고려하자. 

In [4]:
df2 = pd.DataFrame([[None]*20]*100,columns=['커피'+str(i) for i in range(1,11)]+['홍차'+str(i) for i in range(1,11)]) 
df2.index = pd.Index(['user'+str(i) for i in range(1,101)])
df2

커피1   커피2   커피3   커피4   커피5   커피6  ...   홍차5   홍차6   홍차7   홍차8   홍차9  홍차10
user1    None  None  None  None  None  None  ...  None  None  None  None  None  None
user2    None  None  None  None  None  None  ...  None  None  None  None  None  None
user3    None  None  None  None  None  None  ...  None  None  None  None  None  None
user4    None  None  None  None  None  None  ...  None  None  None  None  None  None
user5    None  None  None  None  None  None  ...  None  None  None  None  None  None
...       ...   ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   ...   ...
user96   None  None  None  None  None  None  ...  None  None  None  None  None  None
user97   None  None  None  None  None  None  ...  None  None  None  None  None  None
user98   None  None  None  None  None  None  ...  None  None  None  None  None  None
user99   None  None  None  None  None  None  ...  None  None  None  None  None  None
user100  None  None  None  None  None  None  ...  None  None  None  None  None  None

[100 rows x 20 columns]

In [5]:
for (i,j) in zip(df.user.to_list(), df.item.to_list()):
    df2.iloc[i-1,j-1]=df.query('user == @i and item == @j')['rating'].to_list()[0]

In [6]:
df2.iloc[:5]

커피1      커피2      커피3  ...       홍차8       홍차9      홍차10
user1  4.14921     None     None  ...      None      None  0.903826
user2  4.03181     None     None  ...  0.860954  0.871482      None
user3  4.08218  4.19644     None  ...  0.913116      None  0.956194
user4     None  4.00062  3.89557  ...      None      None   1.14754
user5     None     None     None  ...  0.818332   1.23633      None

[5 rows x 20 columns]

- user1~user5는 대체로 커피를 선호한다. 
- user1은 커피 2,3,7,8,9,10을 먹지는 않았는데, 아마 먹었다면 높은 점수를 주었을 것임. 

In [7]:
df2.iloc[47:53]

커피1       커피2      커피3  ...       홍차8      홍차9      홍차10
user48      None      None  3.71871  ...  0.779323     None      None
user49   4.01202   4.34725   3.7524  ...      None     None      None
user50   4.20859      None  3.99565  ...      None     None  0.985061
user51      None      None  1.15875  ...   4.16822     None      None
user52  0.794371      None     None  ...   4.42073  4.05531      None
user53      None  0.586384  1.17037  ...   3.87398     None      None

[6 rows x 20 columns]

- user1 $\sim$ user50은 커피를 선호하고 user51 $\sim$ user100은 홍차를 선호함 

In [8]:
df2.iloc[94:]

커피1       커피2       커피3  ...      홍차8      홍차9     홍차10
user95    1.19014      None  0.818878  ...   3.9446     None  4.01227
user96   0.511905   1.06614      None  ...     None     None     None
user97       None   1.03502      None  ...     None     None  4.26714
user98       None   1.11551      None  ...  4.11691  3.88782     None
user99       None  0.850794      None  ...     None  3.72529  4.12207
user100      None      None   1.41397  ...  4.10428     None     None

[6 rows x 20 columns]

### 예측? 

`-` None의 값을 추론하는 방법? 머리속으로는 알고있음. 그런데 어떻게 수식화 할지? 

In [9]:
df2.loc['user1']

커피1      4.14921
커피2         None
커피3         None
커피4      4.07814
커피5      4.03342
커피6      4.07187
커피7         None
커피8         None
커피9         None
커피10        None
홍차1      1.14266
홍차2      1.10945
홍차3         None
홍차4     0.603118
홍차5      1.08431
홍차6         None
홍차7     0.906524
홍차8         None
홍차9         None
홍차10    0.903826
Name: user1, dtype: object

`-` 유저의 취향은 예를들면 아래와 같이 표현할 수 있다. 
- user1의 취향 = (커피좋아함, 홍차싫어함) = (0.8, 0.21) 

`-` 코딩해보자.

In [10]:
user1_prprnc = (0.8,0.21) 
user1_prprnc

(0.8, 0.21)

`-` 여기에서 (0.8,0.21)은 각각 (커피음료,홍차음료)와 같은 음료의 특징으로 해석가능. 현재는 단순하게 원소가 2인 벡터이지만 다양한 feature도 가능하다. 
- 음료의특징: (커피음료, 홍차음료, 우유포함, 설탕포함정도, 아이스여부, 행사여부)

`-` 아이템4와 아이템11의 특징은 예를들면 아래와 같이 표현할 수 있다. 

In [11]:
item4_ftr = (0.71, 0.03) # 커피4 
item11_ftr = (0.1, 0.88) # 홍차1  

`-` 유저의 선호도와 아이템의 feature를 내적하여보자. 

In [12]:
np.array(user1_prprnc) @ np.array(item4_ftr) 

0.5742999999999999

In [13]:
np.array(user1_prprnc) @ np.array(item11_ftr) 

0.26480000000000004

`-` 유저의 선호도와 item의 feature가 비슷할수록 값이 높게 나온다. 실제로 user1이 커피4와 홍차1에 매긴 평점은 (4.078139,1.142659) 인데 (0.5743, 0.2648) 가 실제평점과 비슷하게 나오도록 하면 좋겠다. 

`-` 적당히 값을 조정하여 보자. 

In [14]:
user1_prprnc = (0.8,0.21) 
item4_ftr = (5, 0.03) # 커피4 
item11_ftr = (0.1, 4.8) # 홍차1

In [15]:
np.array(user1_prprnc) @ np.array(item4_ftr) 

4.0063

In [16]:
np.array(user1_prprnc) @ np.array(item11_ftr) 

1.0879999999999999

`-` 신기한것은 이 파라메터들이 있다면 None에 대한 추론도 가능하다는 것이다. 예를들어서 user2는 홍차1을 먹지 않았지만 item11_ftr(홍차1)의 `계수값`이 다른유저를 통해 학습되었다면 user2가 홍차1을 먹고 내릴 평점이 예측가능하다. 

In [17]:
df2.loc['user2']

커피1      4.03181
커피2         None
커피3         None
커피4       3.8227
커피5         None
커피6         None
커피7         None
커피8      4.07141
커피9      3.99621
커피10        None
홍차1         None
홍차2     0.839565
홍차3      1.01132
홍차4         None
홍차5      1.12055
홍차6      0.91134
홍차7         None
홍차8     0.860954
홍차9     0.871482
홍차10        None
Name: user2, dtype: object

In [18]:
user2_prprnc = (0.77,0.18) 

In [19]:
np.array(user2_prprnc) @ np.array(item11_ftr)

0.941

`-` 생각해보니까 `user1_prprnc`, ... , `user100_prprnc`, `item1_ftr`, ... , `item20_ftr`를 ${\bf W}$라고 해석하면 우리가 ${\bf W}$ 바꿈에 따라 예측값과 손실이 결정되므로 네트워크 구조가 만들어진다. 

`-` 정리하면 적당한 아래와 같은 매트릭스를 정의하여 

- ${\bf U} = \begin{bmatrix} \tt user1~ prprnc \\ \dots \\ \tt user100~prprnc \end{bmatrix}$ $\quad 100 \times 2$ mat

- ${\bf V} = \begin{bmatrix} \tt item1~ ftr \\ \dots \\ \tt item20~ftr \end{bmatrix}$ $\quad 20 \times 2$ mat

${\bf U}{\bf V}^\top \approx $ `df2` 가 만족하도록 하면 된다. (None은 무시)

### fastai를 학습: df $\to$ dls 

In [20]:
dls=CollabDataLoaders.from_df(df,bs=100) 
# df의 형식: 
#첫번째열은 사용자이 인덱스, 
#두번째열은 아이템의 인덱스, 
#세번쨰열은 rating, 
#네번째열은 아이템의 이름 

In [21]:
dls.items

user  item    rating item_name
296    30    10  4.026810      커피10
840    85     4  1.012898       커피4
512    52     8  0.995410       커피8
283    29     3  4.184027       커피3
596    60    18  3.954355       홍차8
..    ...   ...       ...       ...
312    32     4  4.156176       커피4
570    58     7  0.969826       커피7
238    24     5  3.681099       커피5
984    99     2  0.850794       커피2
704    71    20  3.678341      홍차10

[800 rows x 4 columns]

### fastai를 통한 학습: learn

In [22]:
lrnr = collab_learner(dls,n_factors=2,y_range=(0,5))
lrnr.fit(30,0.01)

In [23]:
lrnr.show_results()

`-` 학습이 잘 되었음. 

### 추천: 마지막유저에 대한 예측결과 (아마 홍차를 좋아할듯)  

In [24]:
X,y = dls.one_batch()

In [25]:
X[:5]

tensor([[22,  2],
        [44,  1],
        [89, 16],
        [71, 20],
        [55, 20]])

In [26]:
y[:5]

tensor([[3.6350],
        [3.9063],
        [3.9940],
        [3.6783],
        [4.1405]])

In [27]:
torch.tensor([[94,3]])

tensor([[94,  3]])

In [28]:
lrnr.model(torch.tensor([[94,3]]).to("cuda:0"))

tensor([1.1823], device='cuda:0', grad_fn=<AddBackward0>)

In [29]:
x100 = torch.tensor([[100,j] for j in range(1,21) ])

In [30]:
lrnr.model(x100.to("cuda:0"))

tensor([0.9848, 1.0528, 1.0282, 1.1276, 1.0393, 0.9778, 1.1069, 1.0857, 1.1415,
        1.2047, 4.0120, 4.1258, 4.0739, 4.1168, 4.0644, 4.1098, 4.0405, 4.1222,
        4.0464, 4.0951], device='cuda:0', grad_fn=<AddBackward0>)

`-` 예상대로 홍차를 매우 좋아함. 

### 숙제 

숙제는 유저2번에 대한 선호도 조사 